runtime/src/comm/ofi/comm-ofi.c

```diff
@@ -4285,9 +4357,154 @@ void amHandler(void* argNil) {
   DBG_PRINTF(DBG_AM, "AM handler done");
 }
 
+static 
+size_t handleAmReq(amRequest_t *req) {
+  size_t size = 0;
+  switch (req->b.op) {
+    case am_opExecOn:
+      DBG_PRINTF(DBG_AM | DBG_AM_RECV, "rx AM sizeof %lu argSize %lu", sizeof(req->xo.hdr), req->xo.hdr.comm.argSize);
+      if (req->xo.hdr.comm.fast) {
+        amWrapExecOnBody(&req->xo.hdr);
+      } else {
+        amHandleExecOn(&req->xo.hdr);
+      }
+      size = req->xo.hdr.comm.argSize;
+      break;
+
+    case am_opExecOnLrg:
+      DBG_PRINTF(DBG_AM | DBG_AM_RECV, "rx AM sizeof %lu argSize %lu", sizeof(req->xol.hdr), req->xol.hdr.comm.argSize);
+      amHandleExecOnLrg(&req->xol.hdr);
+      size = sizeof(req->xol);
+      break;
+
+    case am_opGet:
+      {
+        struct taskArg_RMA_t arg = { .hdr.kind = CHPL_ARG_BUNDLE_KIND_TASK,
+                                     .rma = req->rma, };
+        chpl_task_startMovedTask(FID_NONE, (chpl_fn_p) amWrapGet,
+                                 &arg, sizeof(arg), c_sublocid_any,
+                                 chpl_nullTaskID);
+      }
+      size = sizeof(req->rma);
+      break;
+
+    case am_opPut:
+      {
+        struct taskArg_RMA_t arg = { .hdr.kind = CHPL_ARG_BUNDLE_KIND_TASK,
+                                     .rma = req->rma, };
+        chpl_task_startMovedTask(FID_NONE, (chpl_fn_p) amWrapPut,
+                                 &arg, sizeof(arg), c_sublocid_any,
+                                 chpl_nullTaskID);
+      }
+      size = sizeof(req->rma);
+      break;
+
+    case am_opAMO:
+      amHandleAMO(&req->amo);
+      size = sizeof(req->amo);
+      break;
+
+    case am_opFree:
+      CHPL_FREE(req->free.p);
+      size = sizeof(req->free);
+      break;
+
+    case am_opNop:
+      DBG_PRINTF(DBG_AM | DBG_AM_RECV, "%s", am_reqDoneStr(req));
+      if (req->b.pAmDone != NULL) {
+        amPutDone(req->b.node, req->b.pAmDone);
+      }
+      size = sizeof(req->b);
+      break;
+
+    case am_opShutdown:
+      chpl_signal_shutdown();
+      size = sizeof(req->b);
+      break;
+
+    default:
+      // this isn't recoverable
+      chpl_comm_ofi_abort_on_error = true;
+      INTERNAL_ERROR_V("unexpected AM op %d", (int) req->b.op);
+      break;
+  }
+  return size;
+}
+
+#define MIN(a,b) ((a) < (b) ? (a) : (b))
```

Comment:
I'm always leery of the "simple" form of a construct like this because of the potential that someday someone will invoke it with an actual parameter with side effects which will then happen more than once, but it's your rodeo.  😁

Reply:
Yeah, that is an issue. I've been conditioned not to do that, but maybe not everyone has. I only ended up using the macro twice so I'll remove it.